File: app/src/main/java/com/github/panda_io/micro_panda/ast/statement/if.go
package statement

import (
	"github.com/panda-io/micro-panda/ast/ast"
	"github.com/panda-io/micro-panda/ast/ast_types"
)

type If struct {
	StatementBase
	Initialization ast.Statement
	Condition      ast.Expression
	Body           ast.Statement
	Else           ast.Statement
}

func (i *If) Validate(c ast.Context) {
	ctx := c.NewContext()
	if i.Initialization != nil {
		i.Initialization.Validate(ctx)
	}
	if i.Condition == nil {
		c.Error(i.GetPosition(), "expect condition expression")
	} else {
		i.Condition.Validate(ctx, ast_types.TypeBool)
		if i.Condition.Type() != nil && !i.Condition.Type().Equal(ast_types.TypeBool) {
			c.Error(i.Condition.GetPosition(), "expect bool type condition")
		}
	}
	if i.Body != nil {
		bodyCtx := ctx.NewContext()
		i.Body.Validate(bodyCtx)
	}
	if i.Else != nil {
		elseCtx := ctx.NewContext()
		i.Else.Validate(elseCtx)
	}
}

File: app/src/main/java/com/github/panda_io/micro_panda/ast/statement/If.java
package com.github.panda_io.micro_panda.ast.statement;

import com.github.panda_io.micro_panda.ast.Context;
import com.github.panda_io.micro_panda.ast.type.Type;

public class If extends Statement {
    public Statement initialization;
    public com.github.panda_io.micro_panda.ast.expression.Expression condition;
    public Statement body;
    public Statement elseStatement;

    public void validate(Context context) {
        Context ctx = context.newContext();
        if (this.initialization != null) {
            this.initialization.validate(context);
        }
        if (this.condition == null) {
            context.addError(this.getOffset(), "expect condition expression");
        } else {
            Context conditionCtx = ctx.newContext();
            this.condition.validate(conditionCtx, Type.bool);
            if (this.condition.getType() != null && !this.condition.getType().equal(Type.bool)) {
                context.addError(this.condition.getOffset(), "expect bool type condition");
            }
        }
        if (this.body != null) {
            Context bodyCtx = ctx.newContext();
            this.body.validate(bodyCtx);
        }
        if (this.elseStatement != null) {
            Context elseCtx = ctx.newContext();
            this.elseStatement.validate(elseCtx);
        }
    }
}
